lib/lpc55-rot-startup/src/images.rs

```diff
@@ -56,20 +42,85 @@ extern "C" {
 const PAGE_SIZE: u32 = BYTES_PER_FLASH_PAGE as u32;
 
 pub struct Image {
+    // The boundaries of the flash slot.
     flash: Range<u32>,
-    vector: &'static ImageVectors,
+    // The initial span of programmed flash pages.
+    programmed: Range<u32>,
+    // The number of pages that contain no erased flash words.
+    // _count: usize,
+    // Measurement over the `count`ed pages.
+    fwid: [u8; 32],
 }
 
-pub fn image_details(img: Image, flash: &mut Flash) -> RotImageDetails {
+impl Image {
+    // Before doing any other work with a chunk of flash memory:
+    //
+    //   - Define the address boundaries.
+    //   - Determine the bounds of the initial programmed extent.
+    //   - Count the total number of fully programmed pages.
+    //   - Measure all programmed pages including those outside of the
+    //     initial programmed extent.
+    //
+    // Note: if partially programmed pages is a possibility then that could be
+    // a problem with respect to catching exfiltration attempts.
+    pub fn new(dev: &mut Flash, flash: Range<u32>) -> Image {
+        // let mut _count = 0;
+        let mut end: Option<u32> = None;
+        let mut hash = Sha3_256::new();
+        for start in flash.clone().step_by(BYTES_PER_FLASH_PAGE) {
+            if dev.is_page_range_programmed(start, PAGE_SIZE) {
+                // _count += 1;
+                let page = unsafe {
+                    core::slice::from_raw_parts(
+                        start as *const u8,
+                        BYTES_PER_FLASH_PAGE,
+                    )
+                };
+                hash.update(page);
+            } else if end.is_none() {
+                end = Some(start);
+            }
+        }
+        let fwid = hash.finalize().try_into().unwrap_lite();
+        let programmed = Range {
+            start: flash.start,
+            end: end.unwrap_or(flash.start),
+        };
```

Comment:
Not sure if this is intentional but given a range of 3 blocks with the middle block not programmed the programmed range will be reported as `Range { start: 0, end: 1 }`. Is the intention to ignore all pages after an unprogrammed one is encountered? It will include the 2 programmed page in the hash so the important behavior has been preserved. Representing the programmed pages as a contiguous range when it may not be could cause confusion though.

Reply:
This is intentional. Maybe "programmed" should be renamed for clarity though.
The intent is to know the contiguous programmed region starting from offset zero.
This range will be used to test that pointers and offsets in the NXP executable image have no erased gap pages.
The first test is that the NXP length field is in that region. Later tests are bounded by that
NXP image length value.

Reply:
I updated the comment on Image{programmed: } since I didn't come up with a more descriptive name that wasn't awkward. Does that work?